src/protocol/prf_sharding/mod.rs

```diff
@@ -0,0 +1,704 @@
+        Replicated::share_known_value(&binary_m_ctx, Gf2::ONE),
+        first_row,
+        num_breakdown_key_bits,
+        num_trigger_value_bits,
+        num_saturating_sum_bits,
+    );
+    let mut i: usize = 1;
+    let mut num_users_encountered = 0;
+    let mut row_for_user = 0;
+    while i < input_rows.len() {
+        let cur_row = &input_rows[i];
+        if prev_prf == cur_row.prf_of_match_key {
+            // Do some actual computation
+            let (inputs_required_for_next_row, capped_attribution_outputs) =
+                compute_row_with_previous(
+                    binary_m_ctx.clone(),
+                    RecordId(num_users_encountered),
+                    row_for_user,
+                    cur_row,
+                    &prev_row_inputs,
+                    num_breakdown_key_bits,
+                    num_trigger_value_bits,
+                    num_saturating_sum_bits,
+                )
+                .await?;
+            output.push(capped_attribution_outputs);
+            prev_row_inputs = inputs_required_for_next_row;
+
+            row_for_user += 1;
+        } else {
+            prev_prf = cur_row.prf_of_match_key;
+            prev_row_inputs = initialize_new_device_attribution_variables(
+                Replicated::share_known_value(&binary_m_ctx, Gf2::ONE),
+                cur_row,
+                num_breakdown_key_bits,
+                num_trigger_value_bits,
+                num_saturating_sum_bits,
+            );
+            row_for_user = 0;
+            num_users_encountered += 1;
+        }
+        i += 1;
+    }
+
+    Ok(output)
+}
+
+fn initialize_new_device_attribution_variables<BK, TV>(
+    share_of_one: Replicated<Gf2>,
+    input_row: &PrfShardedIpaInputRow<BK, TV>,
+    num_breakdown_key_bits: usize,
+    num_trigger_value_bits: usize,
+    num_saturating_sum_bits: usize,
+) -> InputsRequiredFromPrevRow
+where
+    BK: GaloisField,
+    TV: GaloisField,
+{
+    InputsRequiredFromPrevRow {
+        ever_encountered_a_source_event: share_of_one - &input_row.is_trigger_bit,
+        attributed_breakdown_key_bits: BitDecomposed::decompose(num_breakdown_key_bits, |i| {
+            Replicated::new(
+                Gf2::truncate_from(input_row.breakdown_key.left()[i]),
+                Gf2::truncate_from(input_row.breakdown_key.right()[i]),
+            )
+        }),
+        saturating_sum: BitDecomposed::new(vec![Replicated::ZERO; num_saturating_sum_bits]),
+        is_saturated: Replicated::ZERO,
+        // This is incorrect in the case that the CAP is less than the maximum value of "trigger value" for a single row
+        // Not a problem if you assume that's an invalid input
+        difference_to_cap: BitDecomposed::new(vec![Replicated::ZERO; num_trigger_value_bits]),
+    }
+}
+
+///
+/// Returns (`sum_bit`, `carry_out`)
+///
+async fn one_bit_adder<C, SB>(
+    ctx: C,
+    record_id: RecordId,
+    x: &SB,
+    y: &SB,
+    carry_in: &SB,
```

Comment:
I guess I don't like mutable anything, so I try to avoid it when possible.

Reply:
I would prefer a struct for the return type rather than a tuple or a mutable argument.

Reply:
it may be a suitable case for a mutable argument here, because the following code is a poor replacement for it imo

```rust
   loop {
       let mut x = 1;
       let y = do_foo(&x);
       x = y;
   }
 ```

in C++ I would also advocate for return struct because it is hard to see that argument is getting mutated, Rust handles this well by forcing you to write `&mut x` at the callee's site.

Reply:
Yeah, now that I've refactored the code into a separate SaturatingSum class, I think the mutable reference is a good solution. I will implement that.